cli_021_snow_cover/contents/src/__init__.py

```diff
@@ -14,328 +14,496 @@
+    # and returns them in HTTP responses.
+    cookie_jar = CookieJar()
+    # return an OpenerDirector instance, which chains the handlers for password_manager and cookie_jar
+    # the OpenerDirector class opens URLs via BaseHandlers chained together
+    opener = urllib.request.build_opener(
+        urllib.request.HTTPBasicAuthHandler(password_manager),
+        urllib.request.HTTPCookieProcessor(cookie_jar))
+    # install an OpenerDirector instance as the default global opener
+    urllib.request.install_opener(opener)
+
+    # make an empty list to store names of the files we downloaded
+    files = []
+    # go through each input date
+    for date in new_dates:
+        # get the url to download the file from the source for the given date
+        url = getUrl(date)
+        # input date is initially a string, strptime changes it to datetime object, strftime reformats into string
+        file_date = datetime.datetime.strptime(date, DATE_FORMAT_HDF).strftime(DATE_FORMAT)
+        # get the filename we want to save the file under locally
+        f = getFilename(file_date)
+        try:
+            # open the url
+            response = urllib.request.urlopen(url)
+            # read the opened url
+            content = response.read()
+            # use BeautifulSoup to read the content as a nested data structure
+            soup = BeautifulSoup(content, 'html.parser')
+
+            hdfs = []
+            # extract all the <a> tags within the content, the goal is to find id of all the hdf files
+            for a in soup.find_all('a'):
+                str_a = str(a)
+                # if an <a> tag contain the word 'hdf'
+                if 'hdf' in str_a:
+                    # find the index of '.hdf'
+                    ext = str_a.index('.hdf')
+                    # slice the id of the .hdf files using the index
+                    hdf = str_a[9:ext+4]
+                    # store all the hdf ids to a list
+                    hdfs.append(hdf)
+            # convert the list to a set to remove duplicates, convert it back to list again
+            hdfs = list(set(hdfs))
+            # get the first item from the list
+            hdf = hdfs[0]
+            # join the source url with the id of each hdf to generate complete URLs
+            # for each hdf file download
+            url = os.path.join(url, hdf)
+
+            try:
+                # try to download the data
+                urllib.request.urlretrieve(url, f)
+                # if successful, add the file to the list of files we have downloaded
+                files.append(f)
+                # if successful, log that the file was downloaded successfully
+                logging.info('Successfully retrieved {}'.format(f))
+
+            except Exception as e:
+                # if unsuccessful, log that the file was not downloaded
+                # (could be because we are attempting to download a file that is not available yet)
```

Comment:
We already checked that this file was on the website, so we know the file is available--something else would have caused the error
```suggestion
```